File: 2017/4-entropy-passphrases/4b.go
package main
/*

Day 4: High-Entropy Passphrases

abcde fghij is a valid passphrase.
abcde xyz ecdab is not valid - the letters from the third word can be
rearranged to form the first word.  a ab abc abd abf abj is a valid passphrase,
because all letters need to be used when forming another word.
iiii oiii ooii oooi oooo is valid.
oiii ioii iioi iiio is not valid - any of these words can be rearranged to form
any other word.

Under this new system policy, how many passphrases are valid?

*/

import (
	"fmt"
    "io/ioutil"
    "sort"
    "strings"
)

func GetChallenge() []string {
    filename := "./input"
    b, _ := ioutil.ReadFile(filename)
    s := strings.TrimSpace(string(b))
    return strings.Split(s, "\n")
}

func Solve(s []string) (valid_count int) {
    for _, v := range s {
        fields := strings.Fields(v)
        words := make(map[string]int8)
        is_valid := bool(true)
        for _, v := range fields {
            runes := strings.Split(v, "")
            sort.Strings(runes)
            sorted_v := strings.Join(runes, "")
            if _, ok := words[sorted_v]; ok {
                is_valid = bool(false)
                break
            } else {
                words[sorted_v] = 1
            }
        }
        if is_valid {
            valid_count++
        }
    }
    return
}

func main() {
    fmt.Println(Solve([]string{
        "abcde fghij",
    }) == 1)
    fmt.Println(Solve([]string{
        "abcde fghij",
        "abcde xyz ecdab",
        "a ab abc abd abf abj",
    }) == 2)
    fmt.Println(Solve(GetChallenge()))
}

File: 2017/4-entropy-passphrases/4b.py
#!/usr/bin/env python3
"""

Day 4: High-Entropy Passphrases

abcde fghij is a valid passphrase.
abcde xyz ecdab is not valid - the letters from the third word can be
rearranged to form the first word.  a ab abc abd abf abj is a valid passphrase,
because all letters need to be used when forming another word.
iiii oiii ooii oooi oooo is valid.
oiii ioii iioi iiio is not valid - any of these words can be rearranged to form
any other word.

Under this new system policy, how many passphrases are valid?

"""


def get_challenge():
    with open("./input", "r") as input_fh:
        return input_fh.readlines()


def solve(puzzle_input):
    valid_count = 0
    for line in puzzle_input:
        line = line.strip()
        if not line:
            continue
        tokens = line.split(" ")
        words = set()
        is_valid = True
        for token in tokens:
            sorted_token = ''.join(sorted([c for c in token]))
            if sorted_token in words:
                is_valid = False
                break
            words.add(sorted_token)
        if is_valid:
            valid_count += 1
    return valid_count


print(solve(
    [
        "abcde fghij",
        ]
    ) == 1)
print(solve(
    [
        "abcde fghij",
        "abcde xyz ecdab",
        "a ab abc abd abf abj",
        ]
    ) == 2)
print(solve(get_challenge()))
